src/video.cpp

```diff
@@ -5,861 +5,908 @@
+		if (max_width < 0) break;
+
+		/* Prevent fuzzy rendering. */
+		x1 = round(x1); y1 = round(y1); x2 = round(x2); y2 = round(y2);
+
+		_video.CoordsToGL(&x1, &y1);
+		_video.CoordsToGL(&x2, &y2);
+
+		GLfloat vertices[6][4] = {
+			{ x1, y2,   0.f, 1.f },
+			{ x2, y1,   1.f, 0.f },
+			{ x1, y1,   0.f, 0.f },
+
+			{ x1, y2,   0.f, 1.f },
+			{ x2, y2,   1.f, 1.f },
+			{ x2, y1,   1.f, 0.f }
+		};
+		glBindTexture(GL_TEXTURE_2D, fg.texture_id);
+		glBindBuffer(GL_ARRAY_BUFFER, this->vbo);
+		glBufferSubData(GL_ARRAY_BUFFER, 0, sizeof(vertices), vertices);
+		glBindBuffer(GL_ARRAY_BUFFER, 0);
+		glDrawArrays(GL_TRIANGLES, 0, 6);
+		x += (fg.advance >> 6) * scale;
 	}
-	return *this;
+	glBindVertexArray(0);
+	glBindTexture(GL_TEXTURE_2D, 0);
 }
 
-/** Initialize the #address if not done already. */
-void ClippedRectangle::ValidateAddress()
+/**
+ * Estimate the bounding rectangle of a string.
+ * @param text Text to estimate.
+ * @param scale Scaling factor for the text size.
+ */
+PointF TextRenderer::EstimateBounds(const std::string &text, float scale) const
 {
-	if (this->address == nullptr) {
-		this->pitch = _video.GetXSize();
-		this->address = _video.mem.get() + this->absx + this->absy * this->pitch;
+	float x = 0;
+	float width = 0;
+	float height = this->font_size * scale;
+	size_t text_length = text.size();
+	for (const char *c = text.c_str(); *c != '\0';) {
+		const FontGlyph &fg = this->GetFontGlyph(&c, text_length);
+		GLfloat xpos = x + fg.bearing.x * scale;
+		GLfloat ypos = (fg.bearing.y - this->characters[BEARING_CHARACTER].bearing.y) * scale;
+		GLfloat w = fg.size.x * scale;
+		GLfloat h = fg.size.y * scale;
+		width = std::max(width, xpos + w);
+		height = std::max(height, ypos + h);
+		x += (fg.advance >> 6) * scale;
 	}
+	return PointF(width + FONT_PADDING_H * this->font_size * scale, height + FONT_PADDING_V * this->font_size * scale);
 }
 
 /**
- * Default constructor, does nothing, never goes wrong.
- * Call #Initialize to initialize the system.
+ * Get the height of a line of text.
+ * @return Height of text.
  */
-VideoSystem::VideoSystem() : initialized(false)
+GLuint TextRenderer::GetTextHeight() const
 {
+	return this->font_size * (FONT_PADDING_V + 1.f);
 }
 
-/** Destructor. */
-VideoSystem::~VideoSystem()
-{
-	if (this->initialized) this->Shutdown();
-}
+/* Graphics framework implementation. */
+
+#ifdef WEBASSEMBLY
+/** Emscripten definitions to query the size of the canvas. */
+EM_JS(int, GetEmscriptenCanvasWidth , (), { return canvas.clientWidth ; });
+EM_JS(int, GetEmscriptenCanvasHeight, (), { return canvas.clientHeight; });
+#endif
 
 /**
- * Initialize the video system, preparing it for use.
- * @param font_name Name of the font file to load.
- * @param font_size Size of the font.
- * @return Error message if initialization failed, else an empty text.
+ * Called by OpenGL when the window size changes.
+ * @param window Window whose size changed.
+ * @param w New width.
+ * @param h New height.
```

Comment:
To make it more clear they are new, change their name. Code like `_video.width = w;` may be interpreted as doing nothing if you miss these two lines.